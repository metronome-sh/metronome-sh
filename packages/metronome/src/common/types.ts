import { ServerBuild } from "@remix-run/server-runtime";
import { type AssetsManifest } from "@remix-run/server-runtime/dist/entry";

export type Routes = Record<string, ServerBuild["routes"][string]>;

export type RouteMap = Record<
  string,
  {
    id: string;
    parentId: string | undefined;
    path: string | undefined;
  }
>;

export type HeaderAllowlist = string[] | "all" | undefined;

export interface MetronomeConfig {
  endpoint?: string | null;
  /**
   * @description You should not use as your API key will be bundled in your server side code. Set `METRONOME_API_KEY` environment variable instead.
   */
  apiKey?: string | null;
  ignoredRoutes?: (string | RegExp)[];
  ignoredPathnames?: (string | RegExp)[];
  headersAllowlist?: HeaderAllowlist;
  debug?: boolean;
}

export interface MetronomeInternalConfig extends Omit<MetronomeConfig, "endpoint"> {
  remixPackages: Record<string, string>;
  endpoint: string;
}

export interface MetronomeWrapperOptions {
  type: "action" | "loader";
  routeId: string;
  routePath?: string;
  config: MetronomeInternalConfig;
  assetsManifest: Pick<AssetsManifest, "version">;
}

export type RegexpRouteMap = Record<
  string,
  {
    id: string;
    parentId: string | undefined;
    path: string | undefined;
    regexp: RegExp;
  }
>;

export type OtelAttribute =
  | string
  | number
  | bigint
  | boolean
  | Array<OtelAttribute>
  | { [key: string]: OtelAttribute };

export type OtelContext = {
  traceId: string;
};

export type AsyncLocalStore = {
  traceId: string;
  requestResolvedAttributes?: Record<string, OtelAttribute>;
};
