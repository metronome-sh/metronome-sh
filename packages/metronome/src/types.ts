import { ServerBuild } from "@remix-run/server-runtime";

export type Routes = Record<string, ServerBuild["routes"][string]>;

export type RouteMap = Record<
  string,
  {
    id: string;
    parentId: string | undefined;
    path: string | undefined;
  }
>;

export type HeaderAllowlist = string[] | "all" | undefined;

export interface MetronomeConfig {
  endpoint?: string | null;
  ignoredRoutes?: (string | RegExp)[];
  ignoredPathnames?: (string | RegExp)[];
  headersAllowlist?: HeaderAllowlist;
}

export interface RemixPackages {
  remixPackages: Record<string, string>;
}

export type MetronomeConfigWithRemixPackages = MetronomeConfig & RemixPackages;

export interface MetronomeWrapperOptions {
  type: "action" | "loader";
  routeId: string;
  version: string;
  routePath?: string;
  config: MetronomeConfigWithRemixPackages;
}
