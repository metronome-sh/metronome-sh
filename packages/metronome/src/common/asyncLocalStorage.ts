import { AsyncLocalStorage } from "async_hooks";
import { AsyncLocalStore } from "./types";

const asyncLocalStorage = new AsyncLocalStorage<AsyncLocalStore>();

export { asyncLocalStorage };
