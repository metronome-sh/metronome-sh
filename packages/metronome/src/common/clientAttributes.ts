import { UAParser } from "ua-parser-js";
import { SemanticAttributes } from "./instrumentation/SemanticAttributes";

export async function getClientAttributes(headers: Headers) {
  const keyValueHeaders = Object.fromEntries(headers.entries());

  const result = await UAParser(keyValueHeaders).withClientHints();

  return {
    [SemanticAttributes.BrowserName]: result.browser.name ?? "",
    [SemanticAttributes.BrowserVersion]: result.browser.version ?? "",
    [SemanticAttributes.BrowserMajor]: result.browser.major ?? "",
    [SemanticAttributes.DeviceModel]: result.device.model ?? "",
    [SemanticAttributes.DeviceVendor]: result.device.vendor ?? "",
    [SemanticAttributes.EngineName]: result.engine.name ?? "",
    [SemanticAttributes.EngineVersion]: result.engine.version ?? "",
    [SemanticAttributes.OsName]: result.os.name ?? "",
    [SemanticAttributes.OsVersion]: result.os.version ?? "",
  };
}
