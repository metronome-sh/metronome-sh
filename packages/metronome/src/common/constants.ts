export const METRONOME_CONTEXT_KEY = "__metronome_context";
export const METRONOME_VERSION = process.env.METRONOME_VERSION!;
export const METRONOME_WEB_VITALS = process.env.METRONOME_WEB_VITALS!;
export const METRONOME_METRICS_VERSION = "v5";
export const METRONOME_REPORT_ROUTE = `/__metronome/${METRONOME_VERSION}/report`;
export const METRONOME_WEB_VITALS_ROUTE = `/__metronome/${METRONOME_VERSION}/web-vitals.js`;
