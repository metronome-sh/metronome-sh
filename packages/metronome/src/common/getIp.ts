// Borrowed from:
// https://github.com/sergiodxa/remix-utils/blob/main/src/server/get-client-ip-address.ts
import { type Request as ExpressRequest } from "express";

const headerNames = [
  "X-Client-IP",
  "X-Forwarded-For",
  "HTTP-X-Forwarded-For",
  "Fly-Client-IP",
  "CF-Connecting-IP",
  "Fastly-Client-Ip",
  "True-Client-Ip",
  "X-Real-IP",
  "X-Cluster-Client-IP",
  "X-Forwarded",
  "Forwarded-For",
  "Forwarded",
  "DO-Connecting-IP" /** Digital ocean app platform */,
  "oxygen-buyer-ip" /** Shopify oxygen platform */,
] as const;

const isIp = (input: string): boolean => {
  const ipRegex =
    /^(?:(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)|(?:[A-Fa-f0-9]{1,4}:){7}[A-Fa-f0-9]{1,4}|(?:[A-Fa-f0-9]{1,4}:){1,7}:|(?:[A-Fa-f0-9]{1,4}:){1,6}:[A-Fa-f0-9]{1,4}|(?:[A-Fa-f0-9]{1,4}:){1,5}(?::[A-Fa-f0-9]{1,4}){1,2}|(?:[A-Fa-f0-9]{1,4}:){1,4}(?::[A-Fa-f0-9]{1,4}){1,3}|(?:[A-Fa-f0-9]{1,4}:){1,3}(?::[A-Fa-f0-9]{1,4}){1,4}|(?:[A-Fa-f0-9]{1,4}:){1,2}(?::[A-Fa-f0-9]{1,4}){1,5}|[A-Fa-f0-9]{1,4}:(?:(?::[A-Fa-f0-9]{1,4}){1,6})|:(?:(?::[A-Fa-f0-9]{1,4}){1,7}|:)|fe80:(?::[A-Fa-f0-9]{0,4}){0,4}%[0-9a-zA-Z]{1,}|::(?:ffff(?::0{1,4}){0,1}:){0,1}(?:(25[0-5]|(2[0-4]|1{0,1}[0-9]){0,1}[0-9])\.){3,3}(25[0-5]|(2[0-4]|1{0,1}[0-9]){0,1}[0-9])|(?:[A-Fa-f0-9]{1,4}:){1,4}:(?:(25[0-5]|(2[0-4]|1{0,1}[0-9]){0,1}[0-9])\.){3,3}(25[0-5]|(2[0-4]|1{0,1}[0-9]){0,1}[0-9]))$/;
  return ipRegex.test(input);
};

function isExpressRequest(
  request: Request | ExpressRequest
): request is ExpressRequest {
  return typeof (request as ExpressRequest).originalUrl !== "undefined";
}

export function getIp(request: Request | ExpressRequest): string | null {
  const headers = request.headers;

  let ipAddress = headerNames
    .flatMap((headerName) => {
      let value: null | string | string[] | undefined;

      if (isExpressRequest(request)) {
        value = request.headers[headerName];
      } else {
        value = request.headers.get(headerName);
      }

      if (headerName === "Forwarded") {
        return parseForwardedHeader(typeof value === "string" ? value : null);
      }

      if (!value) return;

      if (typeof value === "string" && !value.includes(",")) {
        return value;
      }

      return typeof value === "string"
        ? value.split(",").map((ip) => ip.trim())
        : value;
    })
    .find((ip) => {
      if (ip === null || typeof ip === "undefined") return false;
      return isIp(ip);
    });

  return ipAddress ?? null;
}

function parseForwardedHeader(value: string | null): string | null {
  if (!value) return null;
  for (let part of value.split(";")) {
    if (part.startsWith("for=")) return part.slice(4);
    continue;
  }
  return null;
}
