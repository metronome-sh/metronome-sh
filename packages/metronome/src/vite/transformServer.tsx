import { MetronomeResolvedConfig } from "src/common/types";
import MagicString from "magic-string";
import { parse } from "@typescript-eslint/typescript-estree";
import { walk } from "./walk";

export const transformServer = ({
  code,
  id,
  config,
}: {
  code: string;
  id: string;
  config: MetronomeResolvedConfig;
}) => {
  const ast = parse(code, {
    jsx: true,
    useJSXTextNode: true,
    loc: true,
    range: true,
  });

  const magicString = new MagicString(code, { filename: id });

  const excludeFn = config.unstable_exclude?.toString() ?? "undefined";

  const excludePlaceholder = "__exclude_fn__";

  let metronomeConfig = JSON.stringify(
    { ...config, unstable_exclude: excludePlaceholder },
    null,
    2
  );

  metronomeConfig = metronomeConfig.replace(`"${excludePlaceholder}"`, excludeFn);

  magicString.prepend('import { registerMetronome } from "metronome-sh/server";\n');

  walk(ast, (node) => {
    // Find the `const routes` and wrap it
    if (node.type === "VariableDeclaration" && node.declarations.length > 0) {
      node.declarations.forEach((declaration: any) => {
        if (declaration.id && declaration.id.name === "routes" && declaration.init) {
          const [start, end] = declaration.init.range;
          magicString.overwrite(
            start,
            end,
            `registerMetronome(${code.substring(start, end)}, ${metronomeConfig})`
          );
        }
      });
    }
  });

  const serverCode = magicString.toString();

  return {
    code: serverCode,
    map: new MagicString(serverCode).generateMap({
      hires: true,
      includeContent: true,
      source: id,
      file: id,
    }),
  };
};
