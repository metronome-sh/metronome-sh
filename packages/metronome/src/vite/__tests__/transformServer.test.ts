import { describe, expect, it } from "vitest";
import { transformServer } from "../transformServer";
import fs from "fs";
import path from "path";
import { MetronomeResolvedConfig } from "../../common/types";

describe("transformServer", () => {
  it.each(["__templates__/server.txt"])("should transform %s", (filename) => {
    const content = fs.readFileSync(path.resolve(__dirname, filename), "utf-8");

    const config: MetronomeResolvedConfig = {
      apiKey: "test-api-key",
      endpoint: "https://metrics.metronome.sh",
      remixPackages: {
        "remix.package.express": "^2.5.0",
        "remix.package.node": "^2.5.0",
        "remix.package.react": "^2.5.0",
      },
      version: "abcfed",
      unstable_sourceMaps: true,
      unstable_excludeTimeout: 1000,
      unstable_exclude: async () => {
        return true;
      },
    };

    const { code, map } = transformServer({
      code: content,
      id: filename,
      config,
    });

    expect(code).toMatchSnapshot();
    expect(map).toMatchSnapshot();
  });
});
