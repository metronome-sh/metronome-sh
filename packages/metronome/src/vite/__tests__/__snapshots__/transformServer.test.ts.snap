// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`transformServer > should transform __templates__/server.txt 1`] = `
"import { registerMetronome } from "metronome-sh/server";
import { jsx, jsxs } from "react/jsx-runtime";
import { PassThrough } from "node:stream";
import { createReadableStreamFromReadable } from "@remix-run/node";
import { RemixServer, Outlet, Meta, Links, ScrollRestoration, Scripts, useLoaderData } from "@remix-run/react";
import { isbot } from "isbot";
import { renderToPipeableStream } from "react-dom/server";
const ABORT_DELAY = 5e3;
function handleRequest(request, responseStatusCode, responseHeaders, remixContext, loadContext) {
  return isbot(request.headers.get("user-agent") || "") ? handleBotRequest(
    request,
    responseStatusCode,
    responseHeaders,
    remixContext
  ) : handleBrowserRequest(
    request,
    responseStatusCode,
    responseHeaders,
    remixContext
  );
}
function handleBotRequest(request, responseStatusCode, responseHeaders, remixContext) {
  return new Promise((resolve, reject) => {
    let shellRendered = false;
    const { pipe, abort } = renderToPipeableStream(
      /* @__PURE__ */ jsx(
        RemixServer,
        {
          context: remixContext,
          url: request.url,
          abortDelay: ABORT_DELAY
        }
      ),
      {
        onAllReady() {
          shellRendered = true;
          const body = new PassThrough();
          const stream = createReadableStreamFromReadable(body);
          responseHeaders.set("Content-Type", "text/html");
          resolve(
            new Response(stream, {
              headers: responseHeaders,
              status: responseStatusCode
            })
          );
          pipe(body);
        },
        onShellError(error) {
          reject(error);
        },
        onError(error) {
          responseStatusCode = 500;
          if (shellRendered) {
            console.error(error);
          }
        }
      }
    );
    setTimeout(abort, ABORT_DELAY);
  });
}
function handleBrowserRequest(request, responseStatusCode, responseHeaders, remixContext) {
  return new Promise((resolve, reject) => {
    let shellRendered = false;
    const { pipe, abort } = renderToPipeableStream(
      /* @__PURE__ */ jsx(
        RemixServer,
        {
          context: remixContext,
          url: request.url,
          abortDelay: ABORT_DELAY
        }
      ),
      {
        onShellReady() {
          shellRendered = true;
          const body = new PassThrough();
          const stream = createReadableStreamFromReadable(body);
          responseHeaders.set("Content-Type", "text/html");
          resolve(
            new Response(stream, {
              headers: responseHeaders,
              status: responseStatusCode
            })
          );
          pipe(body);
        },
        onShellError(error) {
          reject(error);
        },
        onError(error) {
          responseStatusCode = 500;
          if (shellRendered) {
            console.error(error);
          }
        }
      }
    );
    setTimeout(abort, ABORT_DELAY);
  });
}
const entryServer = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  default: handleRequest
}, Symbol.toStringTag, { value: "Module" }));
function Layout({ children }) {
  return /* @__PURE__ */ jsxs("html", { lang: "en", children: [
    /* @__PURE__ */ jsxs("head", { children: [
      /* @__PURE__ */ jsx("meta", { charSet: "utf-8" }),
      /* @__PURE__ */ jsx("meta", { name: "viewport", content: "width=device-width, initial-scale=1" }),
      /* @__PURE__ */ jsx(Meta, {}),
      /* @__PURE__ */ jsx(Links, {})
    ] }),
    /* @__PURE__ */ jsxs("body", { children: [
      children,
      /* @__PURE__ */ jsx(ScrollRestoration, {}),
      /* @__PURE__ */ jsx(Scripts, {})
    ] })
  ] });
}
function App() {
  return /* @__PURE__ */ jsx(Outlet, {});
}
const route0 = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  Layout,
  default: App
}, Symbol.toStringTag, { value: "Module" }));
const meta = () => {
  return [
    { title: "New Remix App" },
    { name: "description", content: "Welcome to Remix!" }
  ];
};
function loader$1() {
  return {
    message: "Hello from loader!",
    date: /* @__PURE__ */ new Date(),
    promise: Promise.resolve("Hello from a promise!")
  };
}
function Index$1() {
  console.log(useLoaderData());
  return /* @__PURE__ */ jsxs("div", { style: { fontFamily: "system-ui, sans-serif", lineHeight: "1.8" }, children: [
    /* @__PURE__ */ jsx("h1", { children: "Welcome to Remix" }),
    /* @__PURE__ */ jsxs("ul", { children: [
      /* @__PURE__ */ jsx("li", { children: /* @__PURE__ */ jsx(
        "a",
        {
          target: "_blank",
          href: "https://remix.run/tutorials/blog",
          rel: "noreferrer",
          children: "15m Quickstart Blog Tutorial"
        }
      ) }),
      /* @__PURE__ */ jsx("li", { children: /* @__PURE__ */ jsx(
        "a",
        {
          target: "_blank",
          href: "https://remix.run/tutorials/jokes",
          rel: "noreferrer",
          children: "Deep Dive Jokes App Tutorial"
        }
      ) }),
      /* @__PURE__ */ jsx("li", { children: /* @__PURE__ */ jsx("a", { target: "_blank", href: "https://remix.run/docs", rel: "noreferrer", children: "Remix Docs" }) })
    ] })
  ] });
}
const route1 = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  default: Index$1,
  loader: loader$1,
  meta
}, Symbol.toStringTag, { value: "Module" }));
function loader() {
  return {
    message: "Hello from test loader!",
    date: /* @__PURE__ */ new Date(),
    promise: Promise.resolve("Hello from a promise!")
  };
}
function Index() {
  console.log(useLoaderData());
  return /* @__PURE__ */ jsxs("div", { style: { fontFamily: "system-ui, sans-serif", lineHeight: "1.8" }, children: [
    /* @__PURE__ */ jsx("h1", { children: "Welcome to Remix" }),
    /* @__PURE__ */ jsxs("ul", { children: [
      /* @__PURE__ */ jsx("li", { children: /* @__PURE__ */ jsx(
        "a",
        {
          target: "_blank",
          href: "https://remix.run/tutorials/blog",
          rel: "noreferrer",
          children: "15m Quickstart Blog Tutorial"
        }
      ) }),
      /* @__PURE__ */ jsx("li", { children: /* @__PURE__ */ jsx(
        "a",
        {
          target: "_blank",
          href: "https://remix.run/tutorials/jokes",
          rel: "noreferrer",
          children: "Deep Dive Jokes App Tutorial"
        }
      ) }),
      /* @__PURE__ */ jsx("li", { children: /* @__PURE__ */ jsx("a", { target: "_blank", href: "https://remix.run/docs", rel: "noreferrer", children: "Remix Docs" }) })
    ] })
  ] });
}
const route2 = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  default: Index,
  loader
}, Symbol.toStringTag, { value: "Module" }));
const serverManifest = { "entry": { "module": "/assets/entry.client-DxP5pFi7.js", "imports": ["/assets/components-BCN7sZNJ.js"], "css": [] }, "routes": { "root": { "id": "root", "parentId": void 0, "path": "", "index": void 0, "caseSensitive": void 0, "hasAction": false, "hasLoader": false, "hasClientAction": false, "hasClientLoader": false, "hasErrorBoundary": false, "module": "/assets/root-DmlPOGnC.js", "imports": ["/assets/components-BCN7sZNJ.js"], "css": [] }, "routes/_index": { "id": "routes/_index", "parentId": "root", "path": void 0, "index": true, "caseSensitive": void 0, "hasAction": false, "hasLoader": true, "hasClientAction": false, "hasClientLoader": false, "hasErrorBoundary": false, "module": "/assets/_index-C8KVOsa_.js", "imports": ["/assets/components-BCN7sZNJ.js"], "css": [] }, "routes/test": { "id": "routes/test", "parentId": "root", "path": "test", "index": void 0, "caseSensitive": void 0, "hasAction": false, "hasLoader": true, "hasClientAction": false, "hasClientLoader": false, "hasErrorBoundary": false, "module": "/assets/test-C9OOuvRS.js", "imports": ["/assets/components-BCN7sZNJ.js"], "css": [] } }, "url": "/assets/manifest-3fb99c6d.js", "version": "3fb99c6d" };
const mode = "production";
const assetsBuildDirectory = "build/client";
const basename = "/";
const future = { "v3_fetcherPersist": false, "v3_relativeSplatPath": false, "v3_throwAbortReason": false, "unstable_singleFetch": true };
const isSpaMode = false;
const publicPath = "/";
const entry = { module: entryServer };
const routes = registerMetronome({
  "root": {
    id: "root",
    parentId: void 0,
    path: "",
    index: void 0,
    caseSensitive: void 0,
    module: route0
  },
  "routes/_index": {
    id: "routes/_index",
    parentId: "root",
    path: void 0,
    index: true,
    caseSensitive: void 0,
    module: route1
  },
  "routes/test": {
    id: "routes/test",
    parentId: "root",
    path: "test",
    index: void 0,
    caseSensitive: void 0,
    module: route2
  }
}, {
  "apiKey": "test-api-key",
  "endpoint": "https://metrics.metronome.sh",
  "remixPackages": {
    "remix.package.express": "^2.5.0",
    "remix.package.node": "^2.5.0",
    "remix.package.react": "^2.5.0"
  },
  "version": "abcfed",
  "unstable_sourceMaps": true,
  "unstable_excludeTimeout": 1000,
  "unstable_exclude": async () => {
        return true;
      }
});
export {
  serverManifest as assets,
  assetsBuildDirectory,
  basename,
  entry,
  future,
  isSpaMode,
  mode,
  publicPath,
  routes
};
"
`;

exports[`transformServer > should transform __templates__/server.txt 2`] = `
SourceMap {
  "file": "server.txt",
  "mappings": "AAAA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;",
  "names": [],
  "sources": [
    "server.txt",
  ],
  "sourcesContent": [
    "import { registerMetronome } from "metronome-sh/server";
import { jsx, jsxs } from "react/jsx-runtime";
import { PassThrough } from "node:stream";
import { createReadableStreamFromReadable } from "@remix-run/node";
import { RemixServer, Outlet, Meta, Links, ScrollRestoration, Scripts, useLoaderData } from "@remix-run/react";
import { isbot } from "isbot";
import { renderToPipeableStream } from "react-dom/server";
const ABORT_DELAY = 5e3;
function handleRequest(request, responseStatusCode, responseHeaders, remixContext, loadContext) {
  return isbot(request.headers.get("user-agent") || "") ? handleBotRequest(
    request,
    responseStatusCode,
    responseHeaders,
    remixContext
  ) : handleBrowserRequest(
    request,
    responseStatusCode,
    responseHeaders,
    remixContext
  );
}
function handleBotRequest(request, responseStatusCode, responseHeaders, remixContext) {
  return new Promise((resolve, reject) => {
    let shellRendered = false;
    const { pipe, abort } = renderToPipeableStream(
      /* @__PURE__ */ jsx(
        RemixServer,
        {
          context: remixContext,
          url: request.url,
          abortDelay: ABORT_DELAY
        }
      ),
      {
        onAllReady() {
          shellRendered = true;
          const body = new PassThrough();
          const stream = createReadableStreamFromReadable(body);
          responseHeaders.set("Content-Type", "text/html");
          resolve(
            new Response(stream, {
              headers: responseHeaders,
              status: responseStatusCode
            })
          );
          pipe(body);
        },
        onShellError(error) {
          reject(error);
        },
        onError(error) {
          responseStatusCode = 500;
          if (shellRendered) {
            console.error(error);
          }
        }
      }
    );
    setTimeout(abort, ABORT_DELAY);
  });
}
function handleBrowserRequest(request, responseStatusCode, responseHeaders, remixContext) {
  return new Promise((resolve, reject) => {
    let shellRendered = false;
    const { pipe, abort } = renderToPipeableStream(
      /* @__PURE__ */ jsx(
        RemixServer,
        {
          context: remixContext,
          url: request.url,
          abortDelay: ABORT_DELAY
        }
      ),
      {
        onShellReady() {
          shellRendered = true;
          const body = new PassThrough();
          const stream = createReadableStreamFromReadable(body);
          responseHeaders.set("Content-Type", "text/html");
          resolve(
            new Response(stream, {
              headers: responseHeaders,
              status: responseStatusCode
            })
          );
          pipe(body);
        },
        onShellError(error) {
          reject(error);
        },
        onError(error) {
          responseStatusCode = 500;
          if (shellRendered) {
            console.error(error);
          }
        }
      }
    );
    setTimeout(abort, ABORT_DELAY);
  });
}
const entryServer = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  default: handleRequest
}, Symbol.toStringTag, { value: "Module" }));
function Layout({ children }) {
  return /* @__PURE__ */ jsxs("html", { lang: "en", children: [
    /* @__PURE__ */ jsxs("head", { children: [
      /* @__PURE__ */ jsx("meta", { charSet: "utf-8" }),
      /* @__PURE__ */ jsx("meta", { name: "viewport", content: "width=device-width, initial-scale=1" }),
      /* @__PURE__ */ jsx(Meta, {}),
      /* @__PURE__ */ jsx(Links, {})
    ] }),
    /* @__PURE__ */ jsxs("body", { children: [
      children,
      /* @__PURE__ */ jsx(ScrollRestoration, {}),
      /* @__PURE__ */ jsx(Scripts, {})
    ] })
  ] });
}
function App() {
  return /* @__PURE__ */ jsx(Outlet, {});
}
const route0 = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  Layout,
  default: App
}, Symbol.toStringTag, { value: "Module" }));
const meta = () => {
  return [
    { title: "New Remix App" },
    { name: "description", content: "Welcome to Remix!" }
  ];
};
function loader$1() {
  return {
    message: "Hello from loader!",
    date: /* @__PURE__ */ new Date(),
    promise: Promise.resolve("Hello from a promise!")
  };
}
function Index$1() {
  console.log(useLoaderData());
  return /* @__PURE__ */ jsxs("div", { style: { fontFamily: "system-ui, sans-serif", lineHeight: "1.8" }, children: [
    /* @__PURE__ */ jsx("h1", { children: "Welcome to Remix" }),
    /* @__PURE__ */ jsxs("ul", { children: [
      /* @__PURE__ */ jsx("li", { children: /* @__PURE__ */ jsx(
        "a",
        {
          target: "_blank",
          href: "https://remix.run/tutorials/blog",
          rel: "noreferrer",
          children: "15m Quickstart Blog Tutorial"
        }
      ) }),
      /* @__PURE__ */ jsx("li", { children: /* @__PURE__ */ jsx(
        "a",
        {
          target: "_blank",
          href: "https://remix.run/tutorials/jokes",
          rel: "noreferrer",
          children: "Deep Dive Jokes App Tutorial"
        }
      ) }),
      /* @__PURE__ */ jsx("li", { children: /* @__PURE__ */ jsx("a", { target: "_blank", href: "https://remix.run/docs", rel: "noreferrer", children: "Remix Docs" }) })
    ] })
  ] });
}
const route1 = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  default: Index$1,
  loader: loader$1,
  meta
}, Symbol.toStringTag, { value: "Module" }));
function loader() {
  return {
    message: "Hello from test loader!",
    date: /* @__PURE__ */ new Date(),
    promise: Promise.resolve("Hello from a promise!")
  };
}
function Index() {
  console.log(useLoaderData());
  return /* @__PURE__ */ jsxs("div", { style: { fontFamily: "system-ui, sans-serif", lineHeight: "1.8" }, children: [
    /* @__PURE__ */ jsx("h1", { children: "Welcome to Remix" }),
    /* @__PURE__ */ jsxs("ul", { children: [
      /* @__PURE__ */ jsx("li", { children: /* @__PURE__ */ jsx(
        "a",
        {
          target: "_blank",
          href: "https://remix.run/tutorials/blog",
          rel: "noreferrer",
          children: "15m Quickstart Blog Tutorial"
        }
      ) }),
      /* @__PURE__ */ jsx("li", { children: /* @__PURE__ */ jsx(
        "a",
        {
          target: "_blank",
          href: "https://remix.run/tutorials/jokes",
          rel: "noreferrer",
          children: "Deep Dive Jokes App Tutorial"
        }
      ) }),
      /* @__PURE__ */ jsx("li", { children: /* @__PURE__ */ jsx("a", { target: "_blank", href: "https://remix.run/docs", rel: "noreferrer", children: "Remix Docs" }) })
    ] })
  ] });
}
const route2 = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  default: Index,
  loader
}, Symbol.toStringTag, { value: "Module" }));
const serverManifest = { "entry": { "module": "/assets/entry.client-DxP5pFi7.js", "imports": ["/assets/components-BCN7sZNJ.js"], "css": [] }, "routes": { "root": { "id": "root", "parentId": void 0, "path": "", "index": void 0, "caseSensitive": void 0, "hasAction": false, "hasLoader": false, "hasClientAction": false, "hasClientLoader": false, "hasErrorBoundary": false, "module": "/assets/root-DmlPOGnC.js", "imports": ["/assets/components-BCN7sZNJ.js"], "css": [] }, "routes/_index": { "id": "routes/_index", "parentId": "root", "path": void 0, "index": true, "caseSensitive": void 0, "hasAction": false, "hasLoader": true, "hasClientAction": false, "hasClientLoader": false, "hasErrorBoundary": false, "module": "/assets/_index-C8KVOsa_.js", "imports": ["/assets/components-BCN7sZNJ.js"], "css": [] }, "routes/test": { "id": "routes/test", "parentId": "root", "path": "test", "index": void 0, "caseSensitive": void 0, "hasAction": false, "hasLoader": true, "hasClientAction": false, "hasClientLoader": false, "hasErrorBoundary": false, "module": "/assets/test-C9OOuvRS.js", "imports": ["/assets/components-BCN7sZNJ.js"], "css": [] } }, "url": "/assets/manifest-3fb99c6d.js", "version": "3fb99c6d" };
const mode = "production";
const assetsBuildDirectory = "build/client";
const basename = "/";
const future = { "v3_fetcherPersist": false, "v3_relativeSplatPath": false, "v3_throwAbortReason": false, "unstable_singleFetch": true };
const isSpaMode = false;
const publicPath = "/";
const entry = { module: entryServer };
const routes = registerMetronome({
  "root": {
    id: "root",
    parentId: void 0,
    path: "",
    index: void 0,
    caseSensitive: void 0,
    module: route0
  },
  "routes/_index": {
    id: "routes/_index",
    parentId: "root",
    path: void 0,
    index: true,
    caseSensitive: void 0,
    module: route1
  },
  "routes/test": {
    id: "routes/test",
    parentId: "root",
    path: "test",
    index: void 0,
    caseSensitive: void 0,
    module: route2
  }
}, {
  "apiKey": "test-api-key",
  "endpoint": "https://metrics.metronome.sh",
  "remixPackages": {
    "remix.package.express": "^2.5.0",
    "remix.package.node": "^2.5.0",
    "remix.package.react": "^2.5.0"
  },
  "version": "abcfed",
  "unstable_sourceMaps": true,
  "unstable_excludeTimeout": 1000,
  "unstable_exclude": async () => {
        return true;
      }
});
export {
  serverManifest as assets,
  assetsBuildDirectory,
  basename,
  entry,
  future,
  isSpaMode,
  mode,
  publicPath,
  routes
};
",
  ],
  "version": 3,
}
`;
